Linux ci - Compile KVM with goto-cc

nmanthey opened the pull request:
This change adds a github workflow that compiles parts of the Linux kernel with goto-cc, and fails in case compilation fails. Furthermore, the input fails that lead to the failure are archived, to allow easy consumption and investigation.

### Note: This is currently failing, and should help to improve goto-cc to be able to compile Linux

An example failed run with the broken input as artifact has been created as part of my testing: https://github.com/nmanthey/cbmc/pull/5/checks?check_run_id=2093209018

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [N/A] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [N/A] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [N/A] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [N/A] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

nmanthey committed:
linuxci: test compiling linux

Similarly to Xen, CBMC should be able to compile the linux kernel. As
compiling the full kernel is very time consuming, and with CBMC also
space consuming, only compile a core part all its dependencies, with
the smalles configuration possible. For this core part, we select the
KVM hypervisor.

Once this is working, the configuration, as well as targets to be
compiled can be increased.

When the linuxci fails, we want to be able to easily understand the
failure. The used one-line-scan tool already captures input files that
cannot be handled by goto-cc. Hence, also archives these files for easy
access in the web UI.

Signed-off-by: Norbert Manthey <nmanthey@amazon.de>

tautschnig commented on .github/workflows/build-and-test-Linux.yaml:
```diff
@@ -70,5 +70,10 @@ jobs:
       - name: Check for faulty input
         run: ls CPROVER/faultyInput/* || true
 
+      - uses: actions/upload-artifact@v2
+        with:
+          name: CPROVER-faultyInput
+          path: CPROVER/faultyInput/*
```
How does one actually download these? GitHub keeps telling me that I need to have the "actions" scope, perhaps the repository permissions are set up in a way that doesn't permit non-members to do this?

tautschnig replied:
How about doing `tail -n +1 CPROVER/faultyInput/*` instead? Really simple and avoids any permissions problems.

nmanthey replied:
On the checks page (here, e.g.: https://github.com/diffblue/cbmc/pull/5916/checks there is an "artifacts" button on the top left, which allows (me) to download, I did not see any permission restrictions in the description of the action). Yes, I could also dump the content of the files, but that likely floods the output, as those are preprocessed already.

tautschnig replied:
Err, thank you, that was too easy. A first fix is in https://github.com/diffblue/cbmc/pull/5928.

tautschnig commented on .github/workflows/build-and-test-Linux.yaml:
```diff
@@ -0,0 +1,74 @@
+          # user input
+          DEBIAN_FRONTEND: noninteractive
+        run: |
+          sudo apt-get install --no-install-recommends -y coreutils  build-essential gcc git make flex bison software-properties-common curl python
+          sudo apt-get install --no-install-recommends -y bin86 gdb bcc liblzma-dev python-dev gettext iasl uuid-dev libncurses5-dev libncursesw5-dev pkg-config
+          sudo apt-get install --no-install-recommends -y libgtk2.0-dev libyajl-dev sudo time ccache
+          sudo apt-get install --no-install-recommends -y automake bc libssl-dev dkms libelf-dev libudev-dev libpci-dev libiberty-dev autoconf gawk jq
```
As far as I recall, we just need `build-essential libssl-dev` for the Linux kernel.

nmanthey replied:
IIRC, that depends on the configuration. I can minimize this set.

tautschnig commented on .github/workflows/build-and-test-Linux.yaml:
```diff
@@ -0,0 +1,74 @@
+      - name: Get one-line-scan
+        run: git clone -b path-addition https://github.com/awslabs/one-line-scan.git
+
+      - name: Get Linux v5.10
+        run: git clone -b v5.10 --depth=1 https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git linux_5_10
```
I'd suggest to always use the latest release (at the risk of builds surprisingly breaking) and perhaps also using tar/zip instead of cloning?

nmanthey replied:
If people are willing to take the risk, and address the issue (or roll this action back to the last known working release) I can build logic to always access the latest release.

On the other hand, I would like to be able to ensure that a specific version is always working. 

This is the risk I see when always using the latest release: Once the latest release fails, people might ignore the failure or disable this action to unblock themselves. Afterwards, even compiling older relases might start to break again. Hence, I decided to pick a specific release for now.

tautschnig replied:
Maybe we should have two such actions? One with a fixed, longterm-maintenance release (5.10 seems like a reasonable candidate) and one using the "master" branch?

nmanthey replied:
Let's do the stable one for now, and iterate once that one is in.

tautschnig commented:
We will next need #5730 to be merged.

tautschnig commented:
> We will next need #5730 to be merged.

And then #5723, at which point the job should actually succeed.

tautschnig commented:
> > We will next need #5730 to be merged.
> 
> And then #5723, at which point the job should actually succeed.

Both of which are now merged, I believe this test should now succeed.

tautschnig commented on .github/workflows/build-and-test-Linux.yaml:
```diff
@@ -0,0 +1,80 @@
+      - name: Compile Linux with CBMC via one-line-scan, and check for goto-cc section
+        run: |
+          ls
+          make -C linux_5_10 allnoconfig
+          make -C linux_5_10 kvm_guest.config
+          make -C linux_5_10 bzImage -j $(nproc)
```
May I suggest the following variation of this:
```
cat > kvm-config <<EOF
CONFIG_64BIT=y
CONFIG_X86_64=y
CONFIG_HIGH_RES_TIMERS=y
CONFIG_MULTIUSER=y
CONFIG_NET=y
CONFIG_VIRTUALIZATION=y
CONFIG_HYPERVISOR_GUEST=y
CONFIG_PARAVIRT=y
CONFIG_KVM_GUEST=y
CONFIG_KVM=y
CONFIG_KVM_INTEL=y
CONFIG_KVM_AMD=y
EOF
KCONFIG_ALLCONFIG=kvm-config make allnoconfig
make bzImage -j $(nproc)
```

nmanthey replied:
Sure, I can re-iterate

tautschnig commented on .github/workflows/build-and-test-Linux.yaml:
```diff
@@ -0,0 +1,80 @@
+          make -C linux_5_10 kvm_guest.config
+          make -C linux_5_10 bzImage -j $(nproc)
+          du -h . --max-depth=1
+          rm "linux_5_10/arch/x86/kernel/kvm.o"
```
With the above, perhaps instead do:
```
rm arch/x86/kernel/kvm.o
find arch/x86/kvm/ -name "*.o" -delete
```

nmanthey replied:
Will iterate

tautschnig commented on .github/workflows/build-and-test-Linux.yaml:
```diff
@@ -0,0 +1,80 @@
+          make -C linux_5_10 bzImage -j $(nproc)
+          du -h . --max-depth=1
+          rm "linux_5_10/arch/x86/kernel/kvm.o"
+          one-line-scan/one-line-scan --add-to-path $(pwd)/src/cbmc --add-to-path $(pwd)/src/goto-diff  --add-to-path $(pwd)/src/goto-cc --no-analysis --trunc-existing --extra-cflags -Wno-error -o CPROVER -j 3 -- make -C linux_5_10 bzImage -j $(nproc) -k || true
```
I don't think we we should be ignoring failures here, thus drop `-k || true`. Also, shouldn't one-line-scan use (at least) the same `-j` as the command being run?

nmanthey replied:
Yes, that should work.

The '-j' for one-line-scan is limited to not consume too much memory. For github, that should be good enough. For other use cases, that might be higher.

nmanthey committed:
linux: add compilation script

Script the steps required to build linux, so that they can also be
executed locally.

Signed-off-by: Norbert Manthey <nmanthey@amazon.de>

Signed-off-by: Norbert Manthey <nmanthey@amazon.de>

nmanthey commented:
I addressed the comments, and factored out the actual compilation into a separate script, so that these steps can be executed in other environments more easily.

tautschnig commented:
> I addressed the comments, and factored out the actual compilation into a separate script, so that these steps can be executed in other environments more easily.

Thanks a lot, I'll review whether I already have a patch for the invariant failure.

tautschnig commented:
> > I addressed the comments, and factored out the actual compilation into a separate script, so that these steps can be executed in other environments more easily.
> 
> Thanks a lot, I'll review whether I already have a patch for the invariant failure.

A proposal for the patch is in #5967, but I need to reconsider whether that's the correct approach (it is borrowed from #5738, which also made a number of other changes).

NlightNFotis reviewed (commented):
This looks good to me (modulo the CI failure).

I will wait for this to be fixed then I will approve this to get in.

tautschnig commented:
@nmanthey #5967 is now merged, I believe this should pass once rebased.

TGWDB commented on .github/workflows/build-and-test-Linux.yaml:
```diff
@@ -0,0 +1,53 @@
+on:
+  pull_request:
+    branches:
+      - '**'
```
I'm a little hesitant about (another) CI job that runs on all PRs, specially one that we expect to report failures.

nmanthey replied:
Feel free to suggest another branch pattern. I can also target the develop and production branch only. That might result in late surprises.

tautschnig replied:
@TGWDB Why would you expect failures here?

TGWDB replied:
@nmanthey I'm no git expert and don't have a better suggestion. I agree that late failures are not a good idea.

@tautschnig  Based on the original PR text:

"This change adds a github workflow that compiles parts of the Linux kernel with goto-cc, and fails in case compilation fails. Furthermore, the input fails that lead to the failure are archived, to allow easy consumption and investigation."

I'm happy if this works without failures, but was (mis?)reading that we expect this to fail more often due to the target and perhaps related aspects of `cbmc`.

If we expect `goto-cc` to never fail this except for when we break something then I have no objection.
